extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/RestLinksResponseFilter.java

```diff
@@ -0,0 +1,58 @@
+package io.quarkus.resteasy.reactive.links;
+
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Collections;
+
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ResourceInfo;
+import javax.ws.rs.core.Link;
+
+import org.jboss.resteasy.reactive.server.ServerResponseFilter;
+
+public class RestLinksResponseFilter {
+
+    private final RestLinksProvider linksProvider;
+
+    public RestLinksResponseFilter(RestLinksProvider linksProvider) {
+        this.linksProvider = linksProvider;
+    }
+
+    @ServerResponseFilter
+    public void filter(ResourceInfo resourceInfo, ContainerResponseContext responseContext) {
+        for (Link link : getLinks(resourceInfo, responseContext)) {
+            responseContext.getHeaders().add("Link", link);
+        }
+    }
+
+    private Collection<Link> getLinks(ResourceInfo resourceInfo, ContainerResponseContext responseContext) {
+        InjectRestLinks injectRestLinksAnnotation = getInjectRestLinksAnnotation(resourceInfo);
+        if (injectRestLinksAnnotation == null) {
+            return Collections.emptyList();
+        }
+
+        if (injectRestLinksAnnotation.value() == RestLinkType.INSTANCE && responseContext.hasEntity()) {
+            return linksProvider.getInstanceLinks(responseContext.getEntity());
+        }
+
+        return linksProvider.getTypeLinks(getEntityType(resourceInfo.getResourceMethod(), responseContext));
+    }
+
+    private InjectRestLinks getInjectRestLinksAnnotation(ResourceInfo resourceInfo) {
+        if (resourceInfo.getResourceMethod().isAnnotationPresent(InjectRestLinks.class)) {
```

Comment:
Ideally, we want to avoid this kind of thing as much as possible, as it results in Reflection being performed on the method which otherwisr would not have been performed.

One solution to avoiding the reflection for the initial check, is to downcast to `ResteasyReactiveResourceInfo` and use the `getMethodAnnotationNames` method.

You'll probably have to add a similar method for the class check.

Reply:
Another alternative would be to determine the methods that contain the annotations at build time, pass them to a recorder and have your filter look up the information from the info recorded in the recorder.

The general is however that we really want to avoid doing reflection on JAX-RS methods / classes unless it's absolutely necessary